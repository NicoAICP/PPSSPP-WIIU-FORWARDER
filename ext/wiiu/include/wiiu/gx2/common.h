#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
