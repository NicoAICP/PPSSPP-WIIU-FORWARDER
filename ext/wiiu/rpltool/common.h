#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
