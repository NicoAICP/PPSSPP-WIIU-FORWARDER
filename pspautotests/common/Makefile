TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
