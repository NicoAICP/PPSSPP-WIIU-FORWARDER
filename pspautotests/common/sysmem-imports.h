extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();