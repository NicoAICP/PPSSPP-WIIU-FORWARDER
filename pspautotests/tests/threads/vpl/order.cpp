#include <common.h>
#include <pspthreadman.h>

inline void schedVplInfo(SceKernelVplInfo *info) {
	schedf("VPL: OK (size=%d,name=%s,attr=%08X,poolSize=%08X,freeSize=%08X,wait=%d)\n", info->size, info->name, info->attr, info->poolSize, info->freeSize, info->numWaitThreads);
}

inline void schedfVpl(SceUID vpl) {
	if (vpl > 0) {
		SceKernelVplInfo info;
		info.size = sizeof(info);

		int result = sceKernelReferVplStatus(vpl, &info);
		if (result == 0) {
			schedVplInfo(&info);
		} else {
			schedf("VPL: Invalid (%08X)\n", result);
		}
	} else {
		schedf("VPL: Failed (%08X)\n", vpl);
	}
}

struct VplBlock {
	VplBlock *next;
	u32 sizeDiv8;
};

struct VplAccounting {
	void *start;
	void *start2;
	void *startPlusSeven;
	u32 totalSizeMinus8;
	u32 allocatedInBlocks;
	VplBlock *nextFreeBlock;
	VplBlock bottomBlock;
};

void schedfAcct(VplAccounting *acct) {
	if (acct->start != acct->start2) {
		schedf("VPL acct: start %08x, again %08x\n", acct->start, acct->start2);
	}
	if ((char *)acct->start + 7 != (char *)acct->startPlusSeven) {
		schedf("VPL acct: start %08x, startPlusSeven %08x\n", acct->start, acct->startPlusSeven);
	}
	if (acct->nextFreeBlock == 0) {
		schedf("VPL top bad?\n");
	}
	schedf("VPL acct: totalSizeMinus8: %08x, allocated=%08x\n", acct->totalSizeMinus8, acct->allocatedInBlocks * 8);
	schedf("VPL bottom block: at %08x, next->%08x, %08x\n", (char *)&acct->bottomBlock - (char *)acct, (char *)acct->bottomBlock.next - (char *)acct, acct->bottomBlock.sizeDiv8 * 8);

	VplBlock *b = &acct->bottomBlock;
	while (b != NULL) {
		bool isfree = false;
		VplBlock *freeb = acct->nextFreeBlock;
		while (freeb != NULL) {
			if (freeb == b) {
				isfree = true;
				break;
			}
			freeb = freeb->next;
			if (freeb == acct->nextFreeBlock) {
				break;
			}
		}
		const char *desc = "alloc";
		if (acct->nextFreeBlock == b) {
			desc = "next free";
		} else if (isfree) {
			desc = "free";
		}
		schedf("VPL block: at %08x, next->%08x, %08x (%s)\n", (char *)b - (char *)acct, (char *)b->next - (char *)acct, b->sizeDiv8 * 8, desc);
		if (b->sizeDiv8 == 0) {
			break;
		}
		b += b->sizeDiv8;
	}
}

extern "C" int main(int argc, char *argv[]) {
	int result;
	void *addr;
	void *addr1;
	void *addr3;
	void *addrs[50];

	SceUID uid1 = sceKernelCreateVpl("bottom", 2, 0x4300, 0x100, NULL);
	SceUID uid2 = sceKernelCreateVpl("middle", 2, 0x4300, 0x100, NULL);
	SceUID uid3 = sceKernelCreateVpl("top", 2, 0x4300, 0x100, NULL);

	result = sceKernelAllocateVpl(uid1, 0x10, &addr1, NULL);
	result = sceKernelAllocateVpl(uid3, 0x10, &addr3, NULL);

	VplAccounting *acct = (VplAccounting *)((char *)addr3 + 0x18);
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[0], NULL);
	checkpoint("alloc 0");
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[1], NULL);
	checkpoint("alloc 1");
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[2], NULL);
	checkpoint("alloc 2");
	schedfAcct(acct);
	schedfVpl(uid2);

	checkpoint("free 1: %08x", sceKernelFreeVpl(uid2, addrs[1]));
	schedfAcct(acct);
	schedfVpl(uid2);

	checkpoint("free 0: %08x", sceKernelFreeVpl(uid2, addrs[0]));
	schedfAcct(acct);
	schedfVpl(uid2);

	checkpoint("free 2: %08x", sceKernelFreeVpl(uid2, addrs[2]));
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[0], NULL);
	checkpoint("alloc 0");
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[1], NULL);
	checkpoint("alloc 1");
	schedfAcct(acct);
	schedfVpl(uid2);

	checkpoint("free 1: %08x", sceKernelFreeVpl(uid2, addrs[1]));
	schedfAcct(acct);
	schedfVpl(uid2);

	checkpoint("free 0: %08x", sceKernelFreeVpl(uid2, addrs[0]));
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[0], NULL);
	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[1], NULL);
	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[2], NULL);
	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[3], NULL);
	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[4], NULL);
	void *addr2 = addrs[4];
	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[5], NULL);
	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[6], NULL);
	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[7], NULL);
	result = sceKernelAllocateVpl(uid2, 0x18, &addrs[8], NULL);

	schedfAcct(acct);
	schedfVpl(uid2);

	checkpoint("free 4: %08x", sceKernelFreeVpl(uid2, addr2));
	schedfAcct(acct);
	schedfVpl(uid2);

	for (int i = 0; i <= 8; i += 2) {
		sceKernelFreeVpl(uid2, addrs[i]);
		checkpoint("free %d", i);
		schedfAcct(acct);
		schedfVpl(uid2);
	}
	checkpoint("free every other");
	schedfAcct(acct);
	schedfVpl(uid2);

	for (int i = 8; i >= 0; --i) {
		sceKernelFreeVpl(uid2, addrs[i]);
	}
	checkpoint("free all");
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelAllocateVpl(uid2, 0x10, &addrs[0], NULL);
	checkpoint("alloc 1");
	schedfAcct(acct);
	schedfVpl(uid2);

	result = sceKernelFreeVpl(uid2, addrs[0]);
	checkpoint("free 1");
	schedfAcct(acct);
	schedfVpl(uid2);

	return 0;
}
