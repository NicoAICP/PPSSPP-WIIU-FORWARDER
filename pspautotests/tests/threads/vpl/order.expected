VPL acct: totalSizeMinus8: 000000f8, allocated=00000000
VPL bottom block: at 00000018, next->000000f8, 000000e0
VPL block: at 00000018, next->000000f8, 000000e0 (next free)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000E0,wait=0)
[x] alloc 0
VPL acct: totalSizeMinus8: 000000f8, allocated=00000018
VPL bottom block: at 00000018, next->000000f8, 000000c8
VPL block: at 00000018, next->000000f8, 000000c8 (free)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000C8,wait=0)
[x] alloc 1
VPL acct: totalSizeMinus8: 000000f8, allocated=00000030
VPL bottom block: at 00000018, next->000000f8, 000000b0
VPL block: at 00000018, next->000000f8, 000000b0 (free)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000B0,wait=0)
[x] alloc 2
VPL acct: totalSizeMinus8: 000000f8, allocated=00000048
VPL bottom block: at 00000018, next->000000f8, 00000098
VPL block: at 00000018, next->000000f8, 00000098 (free)
VPL block: at 000000b0, next->00000008, 00000018 (alloc)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000098,wait=0)
[x] free 1: 00000000
VPL acct: totalSizeMinus8: 000000f8, allocated=00000030
VPL bottom block: at 00000018, next->000000c8, 00000098
VPL block: at 00000018, next->000000c8, 00000098 (next free)
VPL block: at 000000b0, next->00000008, 00000018 (alloc)
VPL block: at 000000c8, next->000000f8, 00000018 (free)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000B0,wait=0)
[x] free 0: 00000000
VPL acct: totalSizeMinus8: 000000f8, allocated=00000018
VPL bottom block: at 00000018, next->000000c8, 00000098
VPL block: at 00000018, next->000000c8, 00000098 (free)
VPL block: at 000000b0, next->00000008, 00000018 (alloc)
VPL block: at 000000c8, next->000000f8, 00000030 (next free)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000C8,wait=0)
[x] free 2: 00000000
VPL acct: totalSizeMinus8: 000000f8, allocated=00000000
VPL bottom block: at 00000018, next->000000f8, 000000e0
VPL block: at 00000018, next->000000f8, 000000e0 (next free)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000E0,wait=0)
[x] alloc 0
VPL acct: totalSizeMinus8: 000000f8, allocated=00000018
VPL bottom block: at 00000018, next->000000f8, 000000c8
VPL block: at 00000018, next->000000f8, 000000c8 (free)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000C8,wait=0)
[x] alloc 1
VPL acct: totalSizeMinus8: 000000f8, allocated=00000030
VPL bottom block: at 00000018, next->000000f8, 000000b0
VPL block: at 00000018, next->000000f8, 000000b0 (free)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000B0,wait=0)
[x] free 1: 00000000
VPL acct: totalSizeMinus8: 000000f8, allocated=00000018
VPL bottom block: at 00000018, next->000000f8, 000000c8
VPL block: at 00000018, next->000000f8, 000000c8 (next free)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000C8,wait=0)
[x] free 0: 00000000
VPL acct: totalSizeMinus8: 000000f8, allocated=00000000
VPL bottom block: at 00000018, next->000000f8, 000000e0
VPL block: at 00000018, next->000000f8, 000000e0 (next free)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000E0,wait=0)
VPL acct: totalSizeMinus8: 000000f8, allocated=000000e0
VPL bottom block: at 00000018, next->00000008, 00000020
VPL block: at 00000018, next->00000008, 00000020 (alloc)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000008, 00000018 (alloc)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->00000008, 00000018 (alloc)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->00000008, 00000018 (alloc)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->000000f8, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000000,wait=0)
[x] free 4: 00000000
VPL acct: totalSizeMinus8: 000000f8, allocated=000000c8
VPL bottom block: at 00000018, next->00000008, 00000020
VPL block: at 00000018, next->00000008, 00000020 (alloc)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000008, 00000018 (alloc)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->000000f8, 00000018 (free)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->00000008, 00000018 (alloc)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000080, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000018,wait=0)
[x] free 0
VPL acct: totalSizeMinus8: 000000f8, allocated=000000b0
VPL bottom block: at 00000018, next->00000008, 00000020
VPL block: at 00000018, next->00000008, 00000020 (alloc)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000008, 00000018 (alloc)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->000000e0, 00000018 (next free)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->00000008, 00000018 (alloc)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->000000f8, 00000018 (free)
VPL block: at 000000f8, next->00000080, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000030,wait=0)
[x] free 2
VPL acct: totalSizeMinus8: 000000f8, allocated=00000098
VPL bottom block: at 00000018, next->00000008, 00000020
VPL block: at 00000018, next->00000008, 00000020 (alloc)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000008, 00000018 (alloc)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->000000b0, 00000018 (next free)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->000000e0, 00000018 (free)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->000000f8, 00000018 (free)
VPL block: at 000000f8, next->00000080, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000048,wait=0)
[x] free 4
VPL acct: totalSizeMinus8: 000000f8, allocated=00000098
VPL bottom block: at 00000018, next->00000008, 00000020
VPL block: at 00000018, next->00000008, 00000020 (alloc)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000008, 00000018 (alloc)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->000000b0, 00000018 (next free)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->000000e0, 00000018 (free)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->000000f8, 00000018 (free)
VPL block: at 000000f8, next->00000080, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000048,wait=0)
[x] free 6
VPL acct: totalSizeMinus8: 000000f8, allocated=00000080
VPL bottom block: at 00000018, next->00000008, 00000020
VPL block: at 00000018, next->00000008, 00000020 (alloc)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000080, 00000018 (free)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->000000b0, 00000018 (free)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->000000e0, 00000018 (free)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->000000f8, 00000018 (free)
VPL block: at 000000f8, next->00000050, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000060,wait=0)
[x] free 8
VPL acct: totalSizeMinus8: 000000f8, allocated=00000060
VPL bottom block: at 00000018, next->00000050, 00000020
VPL block: at 00000018, next->00000050, 00000020 (free)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000080, 00000018 (free)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->000000b0, 00000018 (free)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->000000e0, 00000018 (free)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->000000f8, 00000018 (free)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000080,wait=0)
[x] free every other
VPL acct: totalSizeMinus8: 000000f8, allocated=00000060
VPL bottom block: at 00000018, next->00000050, 00000020
VPL block: at 00000018, next->00000050, 00000020 (free)
VPL block: at 00000038, next->00000008, 00000018 (alloc)
VPL block: at 00000050, next->00000080, 00000018 (free)
VPL block: at 00000068, next->00000008, 00000018 (alloc)
VPL block: at 00000080, next->000000b0, 00000018 (free)
VPL block: at 00000098, next->00000008, 00000018 (alloc)
VPL block: at 000000b0, next->000000e0, 00000018 (free)
VPL block: at 000000c8, next->00000008, 00000018 (alloc)
VPL block: at 000000e0, next->000000f8, 00000018 (free)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=00000080,wait=0)
[x] free all
VPL acct: totalSizeMinus8: 000000f8, allocated=00000000
VPL bottom block: at 00000018, next->000000f8, 000000e0
VPL block: at 00000018, next->000000f8, 000000e0 (next free)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000E0,wait=0)
[x] alloc 1
VPL acct: totalSizeMinus8: 000000f8, allocated=00000018
VPL bottom block: at 00000018, next->000000f8, 000000c8
VPL block: at 00000018, next->000000f8, 000000c8 (free)
VPL block: at 000000e0, next->00000008, 00000018 (alloc)
VPL block: at 000000f8, next->00000018, 00000000 (next free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000C8,wait=0)
[x] free 1
VPL acct: totalSizeMinus8: 000000f8, allocated=00000000
VPL bottom block: at 00000018, next->000000f8, 000000e0
VPL block: at 00000018, next->000000f8, 000000e0 (next free)
VPL block: at 000000f8, next->00000018, 00000000 (free)
VPL: OK (size=52,name=middle,attr=00004300,poolSize=000000E0,freeSize=000000E0,wait=0)
